plantdescriptionengine/src/main/java/eu/arrowhead/core/plantdescriptionengine/providedservices/pde_mgmt/PlantDescriptionValidator.java

```diff
@@ -0,0 +1,446 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Class for validating Plant Descriptions.
+ */
+public class PlantDescriptionValidator {
+
+    final List<String> blacklist = List.of("unknown");
+    private final List<String> errors = new ArrayList<>();
+
+    /**
+     * Constructor.
+     *
+     * @param entries Plant Description entries to be validated.
+     */
+    public PlantDescriptionValidator(final Map<Integer, ? extends PlantDescriptionEntry> entries) {
+
+        Objects.requireNonNull(entries, "Expected entries.");
+
+        for (PlantDescriptionEntry entry : entries.values()) {
+            validateInIsolation(entry);
+        }
+
+        if (hasError()) {
+            return;
+        }
+
+        // Find the currently active Plant Description, if any.
+        PlantDescriptionEntry activeEntry = entries.values().stream()
+            .filter(PlantDescriptionEntry::active)
+            .findAny()
+            .orElse(null);
+
+        if (activeEntry == null) {
+            return;
+        }
+
+        final Set<PlantDescriptionEntry> activeEntries;
+        try {
+            activeEntries = getIncludeChain(activeEntry, entries);
+        } catch (ValidationException e) {
+            errors.add(e.getMessage());
+            return;
+        }
+
+        ensureIdentifiableSystems(activeEntries);
+        validateConnections(activeEntries);
+    }
+
+    /**
+     * Constructor
+     *
+     * @param entries Plant Description entries to be validated.
+     */
+    public PlantDescriptionValidator(final PlantDescriptionEntry... entries) {
+        this(Arrays.stream(entries)
+            .collect(Collectors.toMap(PlantDescriptionEntry::id, entry -> entry)));
+    }
+
+    /**
+     * Perform all validation checks that can be performed on a Plant
+     * Description entry in isolation.
+     *
+     * @param entry A Plant Description entry to validate.
+     */
+    private void validateInIsolation(final PlantDescriptionEntry entry) {
+        checkSelfReferencing(entry);
+        checkForDuplicateInclusions(entry);
+        validatePorts(entry);
+        validate(entry.systems());
+    }
+
+    private void validate(final List<PdeSystem> systems) {
+        for (final PdeSystem system : systems) {
+            checkIfIdentifiable(system);
+            ensureLegalSystemName(system);
+        }
+    }
+
+    /**
+     * Report an error if the given system does not contain a valid identifier
+     * (either metadata or system name)
+     *
+     * @param system A system to validate.
+     */
+    private void checkIfIdentifiable(final PdeSystem system) {
+        if (system.systemName().isEmpty() && system.metadata().isEmpty()) {
+            errors.add("Contains a system with neither a name nor metadata to identify it.");
+        }
+    }
+
+    /**
+     * Report an error if the given system's name is illegal.
```

Comment:
The method checks the systemId, not the name.

Reply:
Fixed